Add docs for DataStore multi-auth (#3205)
aws-amplify/docs: docs/lib/datastore/fragments/js/setup-auth-rules/multi-auth.md

> ## Multiple Authorization Types

Reviewer: ```suggestion
## Configure Multiple Authorization Types
```

> 

Reviewer: ```suggestion

For some use cases, you will want DataStore to use multiple authorization types. For example, an app might use `API Key` for public content and `Cognito User Pool` for personalized content once the user logs in.

```

> By default, DataStore will use the default authorization type, which is set as the `aws_appsync_authenticationType` key in your Amplify configuration, when synchronizing data with your AWS AppSync API. This means that every network request sent through DataStore will only use that authorization type, no matter what type of `@auth` rule exists on your models.

Reviewer: ```suggestion
By default, DataStore uses your API's default authorization type listed in `aws_appsync_authenticationType` in your Amplify configuration. To change the default authorization type, run `amplify update api`. Every network request sent through DataStore uses that authorization type, regardless of the model's `@auth` rule. 
```

However, if you would like DataStore to infer what type of authorization to use for each model based on the `@auth` rules that exist on a model, you can enable **Multi-Auth** for DataStore by adding the following to your Amplify configuration:

```js
import Amplify, { AuthModeStrategyType } from 'aws-amplify';
import awsconfig from './aws-exports'; 

Amplify.configure({
  ...awsconfig,
  DataStore: {
    authModeStrategyType: AuthModeStrategyType.MULTI_AUTH
  }
})
```

> With this configuration enabled, DataStore will attempt to synchronize data using the `@auth` rule provider for each model.

Reviewer: ```suggestion
This configuration will allow DataStore to synchronize data using the model's `@auth` rule provider for each model.
```

### Multi-Auth priority evaluation

> If there are multiple rules on a model, the rules will be ranked by priority (see below), and DataStore will attempt the synchronization with each authorization type until one succeeds (or they all fail).

Reviewer: ```suggestion
If there are multiple `@auth` rules on a model, the rules will be ranked in priority (see below), and DataStore will attempt the synchronization with each authorization type until one succeeds (or they all fail).
```

> The priority will first be sorted by the `allow` property in the following order:

Reviewer 2: Nit: An ordered list probably fits better here

> The priority will first be sorted by the `allow` property in the following order:

Reviewer: Can't we replace this whole section with:
```
|Priority|`allow`: AuthStrategy|`provider`|
|1|`owner`|`userPools`|
|2|`owner`|`oidc`|
|3|`group`|`userPools`|
|4|`group`|`oidc`|
|5|`private`|`userPools`|
|6|`private`|`iam`|
|7|`pubilc`|`iam`|
|8|`public`|`apiKey`|
```

- `owner`
- `group`
- `private`
- `public`

> If there are multiple rules with the same `allow` value, then the sorting is based on the `provider` property in the following order:

Reviewer 2: Same as above

1. `userPools`
2. `oidc`
3. `iam`
4. `apiKey`

If there is **not** an authenticated user in your application, DataStore will only attempt `public` rules, meaning that `owner`, `group` and `private` rules will be disregarded in order to not waste network requests that we know will fail.

As a fallback, if there are no valid rules for a model, DataStore will attempt synchronization with the default authorization type.

> #### Multi-Auth example

Reviewer: ```suggestion
#### Example with multiple authorization types
```

```graphql
type YourModel
  @model

> 	@auth(

Reviewer 2: ```suggestion
	 @auth(
```

Reviewer 2: It looks like you want to align `@model` and `@auth` (I think :p)

Author: Good observations! I've updated it based on your feedback 👍

		rules: [
			{ allow: owner }
			{ allow: public, provider: apiKey, operations: [read] }
		]
	) {
  ...
}
```

> DataStore will attempt to use owner-based authorization first when synchronizing data if there is an authenticated user. If that request fails for some reason, DataStore will attempt the request again with public authorization. If there is **no** authenticated user, public authorization will be used.

Reviewer: shouldn't we show a DataStore snippet here as well?

Author: Did you have anything particular in mind? The only snippet that would look different would be the configuration, which is already shown above.

To enable DataStore to use multiple authorization types based on the model's `@auth` rules, add `authModeStrategyType: AuthModeStrategyType.MULTI_AUTH` to your Amplify configuration:

```js
import Amplify, { AuthModeStrategyType } from 'aws-amplify';
import awsconfig from './aws-exports'; 

Amplify.configure({
  ...awsconfig,
  DataStore: {
    authModeStrategyType: AuthModeStrategyType.MULTI_AUTH
  }
})
```

This configuration enables DataStore to synchronize data using the model's `@auth` rule provider for each model.

### Multiple authorization types priority order

If there are multiple `@auth` rules on a model, the rules will be ranked by priority (see below), and DataStore will attempt the synchronization with each authorization type until one succeeds (or they all fail).

| Priority | `allow`: AuthStrategy | `provider` |
|:----------|:-----:|:------:|
| 1 | `owner` | `userPools` |
| 2 | `owner` | `oidc` |
| 3 | `group` | `userPools` |
| 4 | `group` | `oidc` |
| 5 | `private` | `userPools` |
| 6 | `private` | `iam` |
| 7 | `public` | `iam` |
| 8 | `public` | `apiKey` |

> If there is **not** an authenticated user in your application, DataStore will only attempt `public` rules, meaning that `owner`, `group` and `private` rules will be disregarded in order to not waste network requests that we know will fail.

Reviewer: ```suggestion
If there is **not** an authenticated user session, DataStore will only attempt `public` rules.
```

shorten for brevity
